### PR TITLE
Limit recursion-depth in dataflow to a configurable constant

### PR DESCRIPTION
**Pull request by hkaiser**

### CODE REVIEW

**Comment by hkaiser** [comment]
This is open for a week now. Please comment asap.

File: src/runtime/threads/threadmanager.cpp [hunk 1]
```diff
@@ -1459,63 +1459,32 @@ namespace hpx { namespace threads
...
         avg_cleanup_idle_rate(bool reset)
     {
-        return avg_creation_idle_rate(reset);
+        return pool_.avg_cleanup_idle_rate(reset);
     }
 #endif
+#endif
 }}
 
 ///////////////////////////////////////////////////////////////////////////////
```
Comment (sithhell) [inline src/runtime/threads/threadmanager.cpp]: The changes here look unrelated to the recursion depth. Can we have them in a different changeset?

File: hpx/lcos/detail/future_data.hpp [hunk 2]
```diff
@@ -219,6 +236,47 @@ namespace detail
             return data_;
         }
 
+        // deferred execution of a given continuation
+        void run_on_completed(completed_callback_type const& on_completed)
```
Comment (sithhell) [inline hpx/lcos/detail/future_data.hpp]: Would it make sense to take the callback by value or rvalue reference here?

**Comment by sithhell** [comment]
The changes to `when_XXX` and `wait_XXX` look unrelated to the issue that this should fix as well.

File: hpx/lcos/detail/future_data.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: What benefit would that give us?

**Comment by hkaiser** [comment]
> The changes to  when_XXX  and  wait_XXX  look unrelated to the issue that this should fix as well.

They are not unrelated. If the futures are deferred, then these changes avoid unlimited recursion inside the when_XXX and wait_XXX functions.

File: src/runtime/threads/threadmanager.cpp (hunk 1 above)
Comment (hkaiser) [inline src/runtime/threads/threadmanager.cpp]: These are unrelated indeed, sorry for those slipping in. They are the result of the necessity to resolve some merge conflicts, things wouldn't compile otherwise.

Comment (hkaiser) [inline src/runtime/threads/threadmanager.cpp]: These changes have been applied to master as well, please disregard them here.

File: hpx/lcos/detail/future_data.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/lcos/detail/future_data.hpp]: Probably keeping constness/non-constness of the passed function. Other than that probably non.

**Comment by sithhell** [comment]
Looks good to me as it seems to fix the issues we currently have. However, those issues might reappear if the limit isn't small enough, for example when a large amount of stack space was previously allocated. I think this needs further thinking instead of trusting on some "magic" number, but the changes are good to be merged

**Comment by hkaiser** [comment]
> Looks good to me as it seems to fix the issues we currently have. However, those issues might reappear if the limit isn't small enough, for example when a large amount of stack space was previously allocated. I think this needs further thinking instead of trusting on some "magic" number, but the changes are good to be merged

I agree, but don't have a better solution at this point.

**Comment by sithhell** [comment]
Reverting this merge as it makes tests failed which are directly related to the changes proposed in this PR. Please fix them before having it on master. Here are the failing tests:
http://hermione.cct.lsu.edu/builders/hpx_appleclang_6_0_boost_1_56_osx_x86_64_debug/builds/236/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_recursion_1613%20%2810.76%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_appleclang_6_0_boost_1_56_osx_x86_64_release/builds/238/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_recursion_1613%20%280.75%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_54_debian_x86_64_debug/builds/525/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_recursion_1613%20%280.28%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_54_debian_x86_64_release/builds/524/steps/run_unit_tests/logs/tests.unit.parallel.service_executors%20%280.36%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_55_debian_x86_64_debug/builds/521/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_recursion_1613%20%281.54%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_55_debian_x86_64_release/builds/496/steps/run_unit_tests/logs/tests.unit.parallel.service_executors%20%280.42%20sec%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_56_debian_x86_64_debug/builds/449/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_launch_775%20%28Timeout%29
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_56_debian_x86_64_debug/builds/449/steps/run_regression_tests/logs/tests.regressions.lcos.dataflow_recursion_1613%20%280.31%20sec%29

and some more.
